Fixes and additional sensor support for Si7006, SHT21, HTU21D sensors

xyzzy42 opened the pull request:
This driver can support other sensors that use the same commands and conversion formulas.

Add SHT21 and HTU21D support.

The math to convert the values was rounding down to an integer.

See individual commits for much more detail.

xyzzy42 committed:
drivers/sensor: si7006: Support SHT21 and HTU21D

These three sensor types are all largely compatible.  The SHT21 and
HTU21D can be supported by this driver by sending command 0xE3 instead
of 0xE0 to read the temperature.

Mention the sensor names in bindings and Kconfig to help those looking
for support to find it.  There have been at least five PRs attempting to
add SHT21 and/or HTU21D support that did not realize the Si7006 is the
same.

As mentioned in PR #22862, the Sensirion SH21 is the original.  The dts
bindings are adjusted (in a backward compatible way!) to make the sht21
the base binding and si7006 is derived from that.

Examples of dts compatibles:

TE Connectivity née Measurement Sepcialties HTU21D:
compatible = "meas,htu21d", "sensirion,sht21";

Sensirion SHT21:
compatible = "sensirion,sht21";

Silicon Labs Si7006
compatible = "silabs,si7006";

Silicon Labs Si7021
compatible = "silabs,si7021", "silabs,si7006";

Signed-off-by: Trent Piepho <tpiepho@gmail.com>

xyzzy42 committed:
drivers/sensor: si7006: Mask off low two bits of data reads

The low two bits are not part of the data, but rather "status" bits that
should be masked off.

This is documented in the HTU21D datasheet Edition 8, date 05/2017, pp.
15, and Sensirion SHT21 datasheet version 6, date 10/2022, §6 (wording
exactly the same):

"The two status bits, the last bits of LSB, must be set to ‘0’ before
calculating physical values."

Also Silicon Labs Si7006 example driver code:
        /* Swap the bytes and clear the status bits */
        return ((data.byte[0] * 256) + data.byte[1]) & ~3;

Since these are the LSBs, it has only a small effect and might not have
been noticed despite being wrong.

While editing this code, switch to using the Zephyr endian conversion
functions intead of a written out conversion.

Add error code to error log message.

Signed-off-by: Trent Piepho <tpiepho@gmail.com>

xyzzy42 committed:
drivers/sensor: si7006: Clean up headers

Some, like stdio.h, don't belong here at all and aren't needed.

Signed-off-by: Trent Piepho <tpiepho@gmail.com>

xyzzy42 committed:
drivers/sensor: si7006: Fix math for calculating sensor values

The existing code rounded the result to an integer, then multiplied that
integer by 1000000 to get micro-degrees or micro-percent, and then
divided by 1000000 to get whole degrees/percent and took the modulus to
get fractional degrees/percent.

Obviously, multiplying and then dividing an integer by the same value
has no effect!

The result is the humidity and temperature were always rounded down to
the nearest integer.

Fix this to properly keep the fractional component.  This is done in a
way that avoids any integer divisions, which are slow on all CPUs, but
especially most microcontrollers, e.g. Cortex-M, lack any integer
division instruction.

Avoiding the base 10 math does not require more code.  One just needs to
think in binary and use binary fractions instead of base 10 fractions.

Signed-off-by: Trent Piepho <tpiepho@gmail.com>

xyzzy42 committed:
drivers/sensor: si7006: Remove unused struct device field

Use was removed but the field itself wasn't deleted.

Signed-off-by: Trent Piepho <tpiepho@gmail.com>

xyzzy42 commented:
This allows the driver to support the three types, as described by @pabigot three years ago back in this comment, https://github.com/zephyrproject-rtos/zephyr/pull/22862#pullrequestreview-359397725

Prior work to try to add these sensors: #11458, #22862, #30185, #38318, #51269.

fkokosinski commented on dts/bindings/sensor/silabs,si7006.yaml:
```diff
@@ -1,8 +1,14 @@
-# Copyright (c) 2019, Electronut Labs
```
Please don't remove old copyright entries.

xyzzy42 replied:
It's a complete re-write of the file.

MaureenHelm replied:
Agree with @fkokosinski

xyzzy42 replied:
Added back.

fkokosinski commented on drivers/sensor/si7006/si7006.c:
```diff
@@ -1,35 +1,30 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#define DT_DRV_COMPAT silabs_si7006
```
Could you extract the generic part of this driver, which would be used by the si7006 and the sht21 drivers? Having it all in one driver named `si7006` makes it confusing.

xyzzy42 replied:
There is just one driver.  The only difference to support both sensors is literally two bits in one byte.

The driver should probably be renamed sht21, since that's the original device.  I can do that in a followup PR.  I find it's much better to do mass-renames in a PR of their own, not combined with a PR that also changes the code in other ways.

MaureenHelm replied:
You need to define `DT_DRV_COMPAT` somewhere for the DT_INST macros below to work. I'd expect to see something like this to support multiple compatibles in the driver: https://github.com/zephyrproject-rtos/zephyr/blob/600e7b7f21cfd179609372bb4f5e3bc5d33d19c2/drivers/sensor/tmp108/tmp108.c#L406-L414

xyzzy42 replied:
I assure you, this driver works.

I don't use the DT_INST macros.  The non "INST" versions work fine.  This seems nicer than needing to undef and redefine DT_DRV_COMPAT between each call.

xyzzy42 replied:
I thought more about splitting the driver.  It doesn't make sense.  tmp108 + as6212 show an example of one driver for two compatible devices.  This is like that.

Splitting it would mean putting everything into the common file.  Then making various static and private data global, i.e. the api definition object, the type for the data and config structs, the init function.  Zephyr doesn't use LTO, so there is a cost to making things non-static.  Then shc21.c and si7006.c become just one line each:
```c
DT_FOREACH_STATUS_OKAY_VARGS(silabs_si7006, SI7006_DEFINE, SI7006_READ_OLD_TEMP);
```
And likewise for sht21.

It's apparent that this just a convoluted way to add more source files solely for the purpose of adding more source files.

And your original issue, that the code for different-but-compatible sensors is in a file that uses the name si7006, is not improved in any way.  All the code (less one line) is still in one file with one name.  

It would be reasonable to mass rename si7006 to something else.  I think sht21, being the original, would be more correct.  Or another name.

MaureenHelm replied:
> I don't use the DT_INST macros. The non "INST" versions work fine. This seems nicer than needing to undef and redefine DT_DRV_COMPAT between each call.

Please keep the INST versions to be consistent with other drivers

xyzzy42 replied:
You can see the same method being used to change a driver from one device to multiple device here, https://github.com/zephyrproject-rtos/zephyr/pull/48042

It seems the existing policy is that method is just fine and no one had any problems with it.

MaureenHelm replied:
There isn't a policy one way or the other for the small number of drivers that support more than one compatible. The vast majority of drivers overall use DT_INST macros and diverging from that pattern makes them harder to review.

xyzzy42 replied:
You demanding that I do something differently, which was acceptable when other authors did it, doesn't feel fair and equitable to me.  If this is important, it would be important for other drivers to do it the same way too.  If it's not important, then why the need to demand it from me?

fabiobaltieri replied:
Hi @xyzzy42, let me try to unstale the conversation here: coding conventions in the repository are global for few things that are well defined in the coding style, but everything else is then down to the subsystem maintainer and collaborators, and ultimately to the code author. It may feel wrong that the author is the lower priority there but the reality is that once the code is accepted, the maintainer is the one responsible for it in the not so uncommon case that the original author becomes unresponsive for whatever reason, so the code has to feel familiar to them.

This case, handling of mutliple compatibles, does not have a well defined pattern in the coding style. If the maintainer and collaborators have no preferences you can pick your style, but in this case they do, so the right thing to do is to follow their preferred style - nothing personal, no hard feelings, they would have asked it to other authors as well if they were submitting code to the subystem they are responsible for.

MaureenHelm reviewed (changes requested):
Please add the new compatible to `tests/drivers/build_all/sensor`

xyzzy42 commented:
> Please add the new compatible to `tests/drivers/build_all/sensor`

Added.  I put it at the end, so the i2c addresses would be used in order.  This seems more useful than putting it near the si7006 binding.

MaureenHelm reviewed (changes requested):
>> Please add the new compatible to tests/drivers/build_all/sensor

> Added. I put it at the end, so the i2c addresses would be used in order. This seems more useful than putting it near the si7006 binding.

Yes it should go at the end, but it doesn't appear that you pushed this change

teburd commented:
Please try to rebase, the failing CI check should go away with a rebase.

xyzzy42 commented:
All failing CI checks are now gone with rebase.

@MaureenHelm, I do not intend to make the change you suggested.  This method is effectively passing implicit parameters to a function via a global variable and I feel that is an anti-pattern to be avoided.  And if anything, the method I have used is more consistent with the existing drivers that support multiple compatible values.

If you feel strongly that what you want is still better, because it makes the drivers with multiple compatibles look more consistent with those with a single compatible, then I offer that demanding I alone change one driver I am writing will not give you this consistency.  New drivers will still use the more common pattern.  And the existing code will still not be written the way you like it.

The existing drivers should be changed to follow the pattern you like too.  And that is clearly out of scope for this PR.

So I propose merging this PR as is, and then as part of the new issue change the existing multi compatible drivers.  That will also give their authors a chance to know about this new design decision.  This way the design of multi-compatible drivers does not need to be decided here in this PR.

xyzzy42 commented:
It has been two months and I can find zero instances of multiple compatible drivers being converted to use the undef DT_DRV_COMPAT pattern.  I guess it's not that important?  Except in this PR, here it is vitally important.  Hard not take that personally.

joseppc commented:
What's bloking this PR?

xyzzy42 commented:
> Contributor

I think the assignee has no interest.

xyzzy42 commented:
I see in PR #56887 that support for multiple compatible properties was originally done using the repeated undef style that @MaureenHelm prefers, but then changed to use the style that I have used in this PR.  And that PR was approved.

So it seems there are others, like @fabiobaltieri, who do not think the repeated undef system is a good design.

Maybe TSC needs to decide?

fabiobaltieri commented:
> So it seems there are others, like @fabiobaltieri, who do not think the repeated undef system is a good design.

Hey hi, I just want to clarify that my comment in that PR (https://github.com/zephyrproject-rtos/zephyr/pull/56887#pullrequestreview-1532403532) was more of a general thought than a request for changes, in fact I wrote that comment while approving the PR. Now, I do thin that that undef/define pattern is awful, but if that's what the subsystem uses and the maintainer is happy so be it.

But that's the point: we don't have a clean and well defined approach to handle driver with multiple compatibles and quirks, it'd be nice to come up with something.

For the PR -- apologies for delay, it probably feel off Maureen radar. Is the multi-driver support format the main contention point? If there's no clear guideline, the priority would be the maintainer preference, they are the ones who have to take care of the code after submission after all.

kartben commented on drivers/sensor/si7006/si7006.c:
```diff
@@ -166,15 +193,17 @@ static int si7006_init(const struct device *dev)
 	return 0;
```
```suggestion
	return -ENOTSUP;
```

kartben replied:
> The way the diff chunks appear makes it look like the "unsupported sensor channel" case, but that is above on line 167 and does return `-NOTSUP`.

Exactly what happened yes - sorry :)

kartben commented on drivers/sensor/si7006/si7006.c:
```diff
@@ -115,25 +109,58 @@ static int si7006_channel_get(const struct device *dev,
 	struct si7006_data *si_data = dev->data;
 
 	if (chan == SENSOR_CHAN_AMBIENT_TEMP) {
```
Might be worthwhile to convert to switch case?

xyzzy42 replied:
I would have used a case statement if I were writing it.  But it was already written by someone else and with only two possibilities the case statement isn't really any better.  I don't want to makes changes just to follow my style better, that just leads to too much code churn.  Everyone has an opinion.  So I'm just trying to change things to make them simpler or faster or fix bugs or add features, etc.

xyzzy42 committed:
drivers/sensor: si7006: Switch to undef DT_DRV_COMPAT

This style is used in sensor drivers.  It's more complicated than the
other way, which is used in different drivers that are not sensor
drivers.

Signed-off-by: Trent Piepho <tpiepho@gmail.com>

xyzzy42 commented:
> > So it seems there are others, like @fabiobaltieri, who do not think the repeated undef system is a good design.
> 
> Hey hi, I just want to clarify that my comment in that PR ([#56887 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/56887#pullrequestreview-1532403532)) was more of a general thought than a request for changes, in fact I wrote that comment while approving the PR. Now, I do think that that undef/define pattern is awful, but if that's what the subsystem uses and the maintainer is happy so be it.
> 
> But that's the point: we don't have a clean and well defined approach to handle driver with multiple compatibles and quirks, it'd be nice to come up with something.
> 
> For the PR -- apologies for delay, it probably feel off Maureen radar. Is the multi-driver support format the main contention point? If there's no clear guideline, the priority would be the maintainer preference, they are the ones who have to take care of the code after submission after all.

I've changed to to the undef/define pattern.  All the other sensors drivers use it.

Of course, one might wonder if having different styles mandated in different types of drivers is really the best policy.  What's next, 4 space indents in some code and 8 space in other code?

I don't think the Linux style exactly of using a table can work.  I see no way to avoid the need to use token pasting to construct the macro names generated from the device tree source.  And there's no way to paste tokens other than a macro invocation.  But a more table like macro could probably be written...

```c
DT_FOREACH_STATUS_OKAY_TABLE(MYDEVICE_DEFINE,
        compat_typea, (0x42, true),
        compat_typeb, (0x10, false),
        compat_typec, (0x00, true)
);
```

Linux is limited to one parameter per compat due to the need to fit into an existing `struct of_device_id` or the like.  Zephyr's static device definition style can allow multiple parameters to be passed to each device.

Also notice how with the DT_INST macro version, the macro that creates the device must add an extra argument, `name`, and then construct the object names with a suffix of `##name##_##inst`.   This is because `inst` starts over from 0 on each invocation and can't be used alone to construct a unique object name.  Clearly a table of compats macro would also want to avoid this and allow constructing object names with an extra argument.

xyzzy42 commented:
Pull request was updated and still awaiting review.

fabiobaltieri commented:
@xyzzy42 could you rebase please?

xyzzy42 commented:
rebased.

Combining the test cases for all sensors into one file causes constant merge conflicts.  Might be better to avoid that by allowing the test to only need to exist in the driver's own files, so there isn't the conflict point of sharing one file with every driver.

fabiobaltieri commented:
@MaureenHelm can you take another look? I think that the main point of contention has been addressed.

margheritamilani commented:
I tried this driver for sht21 sensor and it worked.

str4t0m commented:
With the conflict https://github.com/zephyrproject-rtos/zephyr/pull/73427 might have a higher chance to go in 3.7.0

str4t0m commented:
Rebased on main to resolve the conflict in `tests/drivers/build_all/sensor/i2c.dtsi`

aescolar commented:
@MaureenHelm please re-approve if you think it is good

kartben commented:
@aescolar only change was a rebase to solve conflict, it should be good to go
